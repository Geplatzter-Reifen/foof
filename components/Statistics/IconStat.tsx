import React from "react";
import { StyleSheet, View } from "react-native";
import { Text, Icon, TextProps, useTheme } from "@ui-kitten/components";
import { EvaStatus } from "@ui-kitten/components/devsupport";

type IconStatPropType = React.PropsWithChildren<TextProps> & {
  icon: string;
  iconHeight?: number;
  iconWidth?: number;
  fontSize?: number;
  centered?: boolean;
  reversed?: boolean;
};

export default function IconStat(props: IconStatPropType) {
  const iconHeight = props.iconHeight ?? 22;
  const iconWidth = props.iconWidth ?? "auto";
  const fontSize = props.fontSize ?? 18;
  const centered = props.centered ?? false;
  const reversed = props.reversed ?? false;

  // Iconfarbe wird durch den EvaStatus (oder "text") bestimmt
  const theme = useTheme();
  function getColorStringByStatus(iconStatus: EvaStatus): string {
    switch (iconStatus) {
      case "basic":
        return theme["color-basic-400"];
      case "success":
        return theme["color-success-500"];
      case "info":
        return theme["color-info-500"];
      case "warning":
        return theme["color-warning-500"];
      case "danger":
        return theme["color-danger-500"];
      case "text":
        return theme["text-basic-color"];
      default:
        return theme["color-primary-500"];
    }
  }
  const iconColorString = getColorStringByStatus(props.status ?? "primary");

  return (
    <View
      style={
        centered
          ? styles.viewCentered
          : reversed
            ? styles.viewReversed
            : styles.view
      }
      testID="container"
    >
      <Icon
        name={props.icon}
        style={{
          height: iconHeight,
          width: iconWidth,
          color: iconColorString,
          textAlign: "center",
          textAlignVertical: "center",
        }}
      />
      <Text style={{ fontSize: fontSize, ...styles.text }}>
        {props.children}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  view: {
    flexDirection: "row",
  },
  viewReversed: {
    flex: 1,
    flexDirection: "row-reverse",
  },
  viewCentered: {
    flex: 1,
    alignItems: "center",
    gap: 4,
  },
  text: {
    marginHorizontal: 8,
    flexShrink: 1,
    flexWrap: "wrap",
  },
});
