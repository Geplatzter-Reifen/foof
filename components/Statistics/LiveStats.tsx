import { StyleSheet, View } from "react-native";
import { Icon, Text, useTheme } from "@ui-kitten/components";
import {
  getStageAvgSpeedString,
  getStageDistanceString,
  getStageDurationString,
} from "@/services/statisticsService";
import { Stage } from "@/database/model/model";
import { withObservables } from "@nozbe/watermelondb/react";

function MapStatisticsBox({ stage }: { stage: Stage }) {
  const theme = useTheme();
  const styles = makeStyles(theme);

  if (!stage) {
    return null;
  }

  return (
    <View style={styles.container}>
      <View style={styles.statsRow}>
        <View style={styles.statItem}>
          <Icon name="arrows-left-right" style={styles.icon} />
          <Text style={styles.text}>{getStageDistanceString(stage, 2)}</Text>
        </View>
        <View style={styles.statItem}>
          <Icon name="clock" style={styles.icon} />
          <Text style={styles.text}>{getStageDurationString(stage)}</Text>
        </View>
        <View style={styles.statItem}>
          <Icon name="gauge-high" style={styles.icon} />
          <Text style={styles.text}>{getStageAvgSpeedString(stage)}</Text>
        </View>
      </View>
    </View>
  );
}

// observe the stage (tracks updates to the stage)
const enhance = withObservables(["stage"], ({ stage }: { stage: Stage }) => ({
  stage,
}));

const EnhancedMapStatistics = enhance(MapStatisticsBox);

export default EnhancedMapStatistics;

const makeStyles = (theme) =>
  StyleSheet.create({
    container: {
      position: "absolute",
      top: 650,
      width: "100%",
      padding: 10,
      zIndex: 10,
    },
    statsRow: {
      flexDirection: "row",
      justifyContent: "space-evenly",
      alignItems: "center",
    },
    statItem: {
      flexDirection: "row",
      alignItems: "center",
    },
    icon: {
      height: 20,
      width: 20,
      marginRight: 8,
      color: theme["text-basic-color"],
    },
    text: {
      color: theme["text-basic-color"],
      fontWeight: "bold",
      fontSize: 20,
    },
  });
