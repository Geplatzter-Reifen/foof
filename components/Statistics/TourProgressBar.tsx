import { ProgressBar, Text, ThemeType, useTheme } from "@ui-kitten/components";
import { StyleProp, StyleSheet, View } from "react-native";
import React from "react";

type TourProgressBarProps = React.PropsWithChildren<StyleProp<any>> & {
  progress: number;
};

export function TourProgressBar(props: TourProgressBarProps) {
  const theme = useTheme();
  const styles = makeStyles(theme);

  const { progress, ...rest } = props;

  return (
    <View {...rest}>
      <ProgressBar progress={progress} style={styles.progressBar} />
      <Text
        //@ts-ignore - wegen dem String-Konstrukt bei "left" hat sich TS beschwert
        style={{
          position: "absolute",
          top: "28%",
          left:
            progress < 0.18
              ? `${((progress + 0.04) * 100).toFixed(2)}%`
              : progress < 1
                ? `${((progress - 0.125) * 100).toFixed(2)}%`
                : "44%",
          color: progress >= 0.15 ? "#fff" : theme["text-basic-color"],
          fontSize: 19,
        }}
      >
        {(progress * 100).toFixed(1) + "%"}
      </Text>
    </View>
  );
}

const makeStyles = (theme: ThemeType): any => {
  return StyleSheet.create({
    progressBar: {
      height: 27,
      backgroundColor: theme["background-basic-color-2"],
      borderRadius: 6,
    },
  });
};
