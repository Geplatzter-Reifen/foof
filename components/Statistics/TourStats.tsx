import React from "react";
import { DimensionValue, StyleSheet, View } from "react-native";
import {
  formatDate,
  DateFormat,
  getTotalMillisecondsString,
} from "@/utils/dateUtil";
import {
  Icon,
  ProgressBar,
  Text,
  ThemeType,
  useTheme,
} from "@ui-kitten/components";
import { Tour, Stage } from "@/database/model/model";
import {
  getTourDuration,
  getTourDistance,
  getTourAverageSpeed,
} from "@/services/statisticsService";
import { withObservables } from "@nozbe/watermelondb/react";

let activeTour: Tour | undefined = undefined;
let tourStages: Stage[] = [];

function TourStats() {
  const theme = useTheme();
  const styles = makeStyles(theme);

  const progress = getTourDistance(tourStages) / 1001;

  return (
    <View style={styles.container}>
      <View style={styles.progressContainer}>
        <ProgressBar progress={progress} style={styles.progressBar} />
        <Text
          style={{
            position: "absolute",
            top: "28%",
            left:
              progress < 0.18
                ? (`${((progress + 0.04) * 100).toFixed(2)}%` as DimensionValue)
                : progress < 1
                  ? (`${((progress - 0.125) * 100).toFixed(2)}%` as DimensionValue)
                  : ("44%" as DimensionValue),
            color: progress >= 0.15 ? "#fff" : theme["text-basic-color"],
            fontSize: 19,
          }}
        >
          {(progress * 100).toFixed(1) + "%"}
        </Text>
      </View>
      <View style={styles.statsContainer}>
        <View style={styles.stat_column}>
          <View style={styles.stat_row}>
            <Icon name="arrows-left-right" style={styles.icon_style} />
            <Text>{getTourDistance(tourStages).toFixed(1) + " km"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="arrow-up-right-dots" style={styles.icon_style} />
            <Text>{"0 m"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="gauge-high" style={styles.icon_style} />
            <Text>{getTourAverageSpeed(tourStages).toFixed(1) + " km/h"}</Text>
          </View>
          <View style={styles.stat_row}>
            <Icon name="bolt" style={styles.icon_style} />
            <Text>{0 + " kcal"}</Text>
          </View>
        </View>
        <View style={styles.stat_column}>
          <View style={styles.stat_row}>
            <Icon name="calendar-plus" style={styles.icon_style} />
            <Text>
              {activeTour!.startedAt
                ? formatDate(activeTour!.startedAt, DateFormat.DATE)
                : "--"}
            </Text>
          </View>
          {activeTour!.finishedAt && (
            <View style={styles.stat_row}>
              <Icon name="calendar-check" style={styles.icon_style} />
              <Text>{formatDate(activeTour!.finishedAt, DateFormat.DATE)}</Text>
            </View>
          )}
          <View style={styles.stat_row}>
            <Icon name="clock" style={styles.icon_style} />
            <Text>
              {getTotalMillisecondsString(getTourDuration(tourStages))}
            </Text>
          </View>
        </View>
      </View>
    </View>
  );
}

// observe the stage (tracks updates to the stage)
const enhance = withObservables(["stage"], ({ stage }: { stage: Stage }) => ({
  stage,
}));

const EnhancedTourStats = enhance(TourStats);

// Bridge component that determines which TourStats component to render based on the active stage.
const Bridge = ({ stages }: { stages: Stage[] }) => {
  tourStages = stages;
  for (const stage of stages) {
    if (stage.isActive) {
      return <EnhancedTourStats stage={stage} />;
    }
  }
  return <TourStats />;
};

// observe stages of a tour (only tracks create and delete in the stages table)
const enhanceV2 = withObservables(["tour"], ({ tour }: { tour: Tour }) => {
  activeTour = tour;
  return {
    stages: tour.stages,
  };
});

const EnhancedTourStatsV2 = enhanceV2(Bridge);

export default EnhancedTourStatsV2;

const makeStyles = (theme: ThemeType) => {
  return StyleSheet.create({
    container: {
      backgroundColor: theme["color-primary-transparent-500"],
    },
    progressContainer: {
      padding: 8,
      margin: 10,
      marginBottom: 0,
    },
    progressBar: {
      height: 27,
      //@ts-ignore
      indicatorColor: theme["color-primary-500"],
      backgroundColor: theme["background-basic-color-2"],
      borderRadius: 6,
    },
    statsContainer: {
      flexDirection: "row",
      alignItems: "center",
      height: "auto",
      paddingHorizontal: 20,
      paddingBottom: 10,
    },
    stat_column: {
      flex: 1,
      alignItems: "flex-start",
      justifyContent: "flex-start",
    },
    stat_row: {
      flexDirection: "row",
      alignItems: "center",
    },
    icon_style: {
      alignSelf: "center",
      marginHorizontal: 4,
      height: 15,
      width: "100%",
      minWidth: 17,
    },
  });
};
