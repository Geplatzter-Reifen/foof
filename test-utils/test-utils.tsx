// https://testing-library.com/docs/react-testing-library/setup/#add-custom-queries
import React from "react";
import { render, RenderOptions } from "@testing-library/react-native";
import { FontAwesomeIconsPack } from "@/components/Font/fontAwesome";
import { mapping } from "@eva-design/eva";
import {
  IconRegistry,
  ApplicationProvider,
  ModalService,
} from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { foofDarkTheme, foofLightTheme } from "@/constants/custom-theme";
import { library } from "@fortawesome/fontawesome-svg-core";
import { far } from "@fortawesome/free-regular-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { fab } from "@fortawesome/free-brands-svg-icons";

// Iconpacks laden
library.add(far, fas, fab);

// ModalService wird verwendet, um die Modal-Positionierung zu steuern
ModalService.setShouldUseTopInsets = true;

type Props = {
  children: React.ReactNode;
  use_dark_theme?: boolean;
};
// Der Wrapper wird verwendet, um die Provider zu umschließen, die für die Komponenten notwendig sind
const AllTheProviders = ({ children, use_dark_theme = true }: Props) => {
  const theme = use_dark_theme
    ? { ...eva.dark, ...foofDarkTheme }
    : { ...eva.light, ...foofLightTheme };
  return (
    <>
      <IconRegistry icons={FontAwesomeIconsPack} />
      <ApplicationProvider {...eva} customMapping={mapping} theme={theme}>
        {children}
      </ApplicationProvider>
    </>
  );
};

// Erstellt einen Custom-Render, der die Provider um die Komponenten wickelt
const customRender = (
  component: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(component, { wrapper: AllTheProviders, ...options });
};

// Exportiert alle Funktionen der Testing Library, weswegen man diese dann auch von dieser Datei importieren kann
export * from "@testing-library/react-native";
export { customRender as render };
