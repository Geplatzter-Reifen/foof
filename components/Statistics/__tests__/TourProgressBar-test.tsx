import { render } from "@/test-utils/test-utils";
import { TourProgressBar } from "@/components/Statistics/TourProgressBar";

describe("TourProgressBar", () => {
  it("renders correctly for 0 Percent", async () => {
    const bar = render(<TourProgressBar progress={0} />);
    expect(bar).toMatchSnapshot();
  });
  it("renders correctly for < 18% Percent", async () => {
    const bar = render(<TourProgressBar progress={0.175} />);
    expect(bar).toMatchSnapshot();
  });
  it("renders correctly for 18% Percent", async () => {
    const bar = render(<TourProgressBar progress={0.18} />);
    expect(bar).toMatchSnapshot();
  });
  it("renders correctly for > 18% Percent", async () => {
    const bar = render(<TourProgressBar progress={0.543} />);
    expect(bar).toMatchSnapshot();
  });
  it("renders correctly for 100% Percent", async () => {
    const bar = render(<TourProgressBar progress={1} />);
    expect(bar).toMatchSnapshot();
  });
  it("renders correctly for > 100% Percent", async () => {
    const bar = render(<TourProgressBar progress={1.234} />);
    expect(bar).toMatchSnapshot();
  });
});
