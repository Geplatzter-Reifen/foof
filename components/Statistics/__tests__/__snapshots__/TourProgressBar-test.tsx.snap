// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourProgressBar renders correctly for < 18% Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "17.5%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#000000",
          "fontSize": 19,
          "left": "21.50%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    17.5%
  </Text>
</View>
`;

exports[`TourProgressBar renders correctly for > 18% Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "54.300000000000004%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#fff",
          "fontSize": 19,
          "left": "41.80%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    54.3%
  </Text>
</View>
`;

exports[`TourProgressBar renders correctly for > 100% Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "100%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#fff",
          "fontSize": 19,
          "left": "44%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    123.4%
  </Text>
</View>
`;

exports[`TourProgressBar renders correctly for 0 Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "0%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#000000",
          "fontSize": 19,
          "left": "4.00%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    0.0%
  </Text>
</View>
`;

exports[`TourProgressBar renders correctly for 18% Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "18%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#fff",
          "fontSize": 19,
          "left": "5.50%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    18.0%
  </Text>
</View>
`;

exports[`TourProgressBar renders correctly for 100% Percent 1`] = `
<View>
  <View
    appearance="default"
    onLayout={[Function]}
    size="small"
    status="primary"
    style={
      [
        {
          "backgroundColor": "rgba(51, 102, 255, 0.24)",
          "borderRadius": 6,
          "height": 27,
        },
        {
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#eaeaea",
          "borderRadius": 6,
          "height": 27,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#FF6E14",
          "borderRadius": 6,
          "height": 27,
          "width": "100%",
        }
      }
    />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "color": "#fff",
          "fontSize": 19,
          "left": "44%",
          "position": "absolute",
          "top": "28%",
        },
      ]
    }
  >
    100.0%
  </Text>
</View>
`;
